fendermint/eth/api/src/mpool.rs

```diff
@@ -25,22 +27,81 @@ const RETRY_SLEEP_SECS: u64 = 5;
 /// being dropped from the mempool.
 pub type TransactionCache = Cache<et::TxHash, et::Transaction>;
 
-pub fn start_tx_cache_clearing(client: FendermintClient<HybridClient>, cache: TransactionCache) {
+/// Buffer out-of-order messages until they can be sent to the chain.
+#[derive(Clone)]
+pub struct TransactionBuffer(pub Cache<Address, BTreeMap<Nonce, ChainMessage>>);
+
+impl TransactionBuffer {
+    /// Insert a transaction we could not submit straight away into the buffer.
+    pub fn insert(&self, sender: Address, nonce: Nonce, msg: ChainMessage) {
+        self.0.with(|c| {
+            let buffer = c.entry(sender).or_insert_with(BTreeMap::new);
+            // Overwrite any previous entry to protect against DoS attack; it wouldn't make sense to submit them anyway.
+            buffer.insert(nonce, msg);
+        })
+    }
+
+    /// Remove all (sender, nonce) pairs which were included in a block.
+    fn remove_many<'a, I>(&self, txs: I)
+    where
+        I: Iterator<Item = (&'a Address, Nonce)>,
+    {
+        self.0.with(|c| {
+            for (sender, nonce) in txs {
+                if let Some(buffer) = c.get_mut(sender) {
+                    buffer.remove(&nonce);
+                }
+            }
+        })
+    }
+
+    /// Gather any messages that have been enabled by transactions added to a block.
+    ///
+    /// These are removed from the cache, submission is only attempted once.
+    fn remove_enabled<'a, I>(&self, txs: I) -> Vec<(Address, Nonce, ChainMessage)>
+    where
+        I: Iterator<Item = (&'a Address, Nonce)>,
+    {
+        self.0.with(|c| {
+            let mut msgs = Vec::new();
+            for (sender, mut nonce) in txs {
+                if let Some(buffer) = c.get_mut(sender) {
+                    nonce += 1;
+                    while let Some(msg) = buffer.remove(&nonce) {
```

Comment:
this is more like a depth first removal, not sure if this is fair to other addresses that might have submitted the txn earlier. Actually is fairness a concern at all here.

Reply:
You are right it is depth first. This was the simplest code I could come up with, which I believe is fine for a first PR. If you have a more elaborate strategy then we can create a ticket for it. 

I think ultimately fairness should be built into the proposal methods, as you won't be able to control how transactions are submitted to the mempool, but there you can order them to your liking.